File: packages/contracts/python/afrosite_contracts/guardrails.py
"""Garde-fous de génération — lit le même JSON que la version TypeScript."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from functools import cache
from pathlib import Path

GUARDRAILS_FILE = (
    Path(__file__).resolve().parents[2] / "generation-guardrails.json"
)


@dataclass(frozen=True)
class QualityCheck:
    passed: bool
    errors: tuple[str, ...]


@cache
def _guardrails() -> dict:
    return json.loads(GUARDRAILS_FILE.read_text(encoding="utf-8"))


def guardrails_version() -> str:
    return _guardrails()["schema_version"]


def scan_generated_artifact(
    *, files: tuple[str, ...] | list[str], diff: str
) -> QualityCheck:
    rules = _guardrails()
    errors: list[str] = []
    denied_path = re.compile(rules["denied_path_pattern"], re.IGNORECASE)
    for path in files:
        if denied_path.search(path.replace("\\", "/")):
            errors.append(f"{rules['denied_path_message']}: {path}")
    for entry in rules["denied_diff_patterns"]:
        if re.search(entry["pattern"], diff, re.IGNORECASE):
            errors.append(entry["message"])
    return QualityCheck(passed=not errors, errors=tuple(errors))

File: packages/contracts/src/guardrails.ts
import guardrails from "../generation-guardrails.json";

export type GuardrailPattern = {
  id: string;
  pattern: string;
  message: string;
};

export const GENERATION_GUARDRAILS: {
  schema_version: string;
  denied_path_pattern: string;
  denied_path_message: string;
  denied_diff_patterns: GuardrailPattern[];
} = guardrails;

export function deniedPathMatcher(): RegExp {
  return new RegExp(GENERATION_GUARDRAILS.denied_path_pattern, "i");
}

export function deniedDiffMatchers(): (GuardrailPattern & { regex: RegExp })[] {
  return GENERATION_GUARDRAILS.denied_diff_patterns.map((entry) => ({
    ...entry,
    regex: new RegExp(entry.pattern, "i"),
  }));
}
